### PR TITLE
Add nodeinfo to iamclient

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: fcrypt/fcrypt.go [hunk 1]
```diff
@@ -112,7 +111,7 @@ type SignContext struct {
 type SignContextInterface interface {
 	AddCertificate(fingerprint string, asn1Bytes []byte) (err error)
 	AddCertificateChain(name string, fingerprints []string) (err error)
-	VerifySign(ctx context.Context, f *os.File, sign *cloudprotocol.Signs) (err error)
+	// VerifySign(ctx context.Context, f *os.File, sign *cloudprotocol.Signs) (err error)
```
Comment (al1img) [inline fcrypt/fcrypt.go:114]: Drop this commit: do not comment out

File: iamclient/iamclient.go [hunk 2]
```diff
@@ -78,6 +85,15 @@ type CertificateProvider interface {
 	GetCertSerial(certURL string) (serial string, err error)
 }
 
+// NodeInfoProvider provides information about the nodes.
```
Comment (al1img) [inline iamclient/iamclient.go:88]: Interface should be defined in place where it used. Remove it.

File: iamclient/iamclient.go [hunk 3]
```diff
@@ -141,6 +162,51 @@ func (client *Client) GetSystemID() (systemID string) {
 	return client.systemID
 }
 
+// GetNodeInfo returns node info.
+func (client *Client) GetNodeInfo(nodeID string) (nodeInfo *pb.NodeInfo, err error) {
```
Comment (al1img) [inline iamclient/iamclient.go:166]: No PB types in public API. It should return `cloudprotocol.NodeInfo`. I will add conversion function for aos_common pvconvert.

File: iamclient/iamclient.go [hunk 4]
```diff
@@ -141,6 +162,51 @@ func (client *Client) GetSystemID() (systemID string) {
...
+	response, err := client.nodeService.GetNodeInfo(ctx, request)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	return response, err
+}
+
+// GetAllNodeIDs returns node ids.
+func (client *Client) GetAllNodeIDs() (nodesId *pb.NodesID, err error) {
```
Comment (al1img) [inline iamclient/iamclient.go:183]: Ditto. Should return `[]string`

File: iamclient/iamclient.go [hunk 5]
```diff
@@ -141,6 +162,51 @@ func (client *Client) GetSystemID() (systemID string) {
...
+	response, err := client.nodeService.GetAllNodeIDs(ctx, &emptypb.Empty{})
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	return response, err
+}
+
+// SubscribeNodeInfoChange subscribes client on NodeInfoChange events
+func (client *Client) SubscribeNodeInfoChange() <-chan *pb.NodeInfo {
```
Comment (al1img) [inline iamclient/iamclient.go:198]: Ditto. `cloudprotocol.NodeInfo`

File: fcrypt/fcrypt.go (hunk 1 above)
Comment (mykola-kobets-epam) [reply fcrypt/fcrypt.go]: done

File: iamclient/iamclient.go (hunk 2 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: done

File: iamclient/iamclient.go (hunk 3 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: done

File: iamclient/iamclient.go (hunk 4 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go]: done

File: iamclient/iamclient.go [hunk 6]
```diff
@@ -119,13 +130,17 @@ func New(
...
 	if localClient.nodeID, _, err = localClient.getNodeInfo(); err != nil {
-		return client, aoserrors.Wrap(err)
+		return nil, aoserrors.Wrap(err)
 	}
 
 	if localClient.systemID, err = localClient.getSystemID(); err != nil {
 		return nil, aoserrors.Wrap(err)
 	}
 
+	localClient.nodeInfoListeners = []chan cloudprotocol.NodeInfo{}
```
Comment (al1img) [inline iamclient/iamclient.go:140]: use make in `localclient` initializaion

File: iamclient/iamclient.go [hunk 7]
```diff
@@ -139,6 +154,51 @@ func (client *Client) GetSystemID() (systemID string) {
...
+
+	return response.Ids, err
+}
+
+// SubscribeNodeInfoChange subscribes client on NodeInfoChange events
+func (client *Client) SubscribeNodeInfoChange() <-chan cloudprotocol.NodeInfo {
+	client.Lock()
+	defer client.Unlock()
+
+	log.Debug("Subscribe on NodeInfoChanged event")
```
Comment (al1img) [inline iamclient/iamclient.go:194]: Do not use function name or variable names in log

File: iamclient/iamclient.go [hunk 8]
```diff
@@ -356,3 +418,73 @@ func (client *Client) getSystemID() (systemID string, err error) {
...
+		client.Lock()
+		for _, listener := range client.nodeInfoListeners {
+			listener <- pbconvert.NewNodeInfoFromPB(nodeInfo)
+		}
+		client.Unlock()
+	}
+}
+
+func (client *Client) connectPublicNodeService() {
+	go func() {
```
Comment (al1img) [inline iamclient/iamclient.go:445]: Better to use `go connectPublicNodeService()` to decrease function nested level.

File: iamclient/iamclient.go (hunk 6 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:140]: fixed

File: iamclient/iamclient.go (hunk 7 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:194]: done

File: iamclient/iamclient.go (hunk 8 above)
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:445]: done

File: iamclient/iamclient.go [hunk 9]
```diff
@@ -139,6 +154,51 @@ func (client *Client) GetSystemID() (systemID string) {
...
+	response, err := client.nodeService.GetNodeInfo(ctx, request)
+	if err != nil {
+		return cloudprotocol.NodeInfo{}, aoserrors.Wrap(err)
+	}
+
+	return pbconvert.NewNodeInfoFromPB(response), nil
+}
+
+// GetAllNodeIDs returns node ids.
+func (client *Client) GetAllNodeIDs() (nodesId []string, err error) {
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply iamclient/iamclient.go:175]: done

**Review by al1img — APPROVED** [review]

**Pull request merged**